File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    // have two stacks (left , right)
    // 1. left - insert in ascending
    // 2. right - insert in descending
    // 3.  do two sum based on comparing the values  like we did for
    // vectors 
    void insertAsc(TreeNode* node,stack<TreeNode*>& stk){
        while(node){
            stk.push(node);
            node=node->left;
        }
    }

    void insertDesc(TreeNode* node,stack<TreeNode*>& stk){
        while(node){
            stk.push(node);
            node=node->right;
        }
    }

    bool findTarget(TreeNode* root, int k) {
        stack<TreeNode*> left, right; 
        insertAsc(root,left);
        insertDesc(root,right);
        while(left.size() && right.size() && left.top()!=right.top()){
            int l=left.top()->val;
            int r=right.top()->val;
            // if(l>root->val || r<=root->val) return false;
            if(l+r==k && left.top()!=right.top()) return true;
            if(l+r>k){
                TreeNode* node = right.top();
                right.pop();
                insertDesc(node->left,right);
            }
            else {
                TreeNode* node = left.top();
                left.pop();
                insertAsc(node->right,left);
            }
        }
        return false;
    }
};

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def nextItr(self,st):
        top = st.pop() 
        node = top.right 
        while node : 
            st.append(node)
            node=node.left 
        return st 
    
    def prevItr(self,st): 
        top = st.pop() 
        node = top.left 
        while node: 
            st.append(node)
            node=node.right 
        return st 

    def findTarget(self, root: Optional[TreeNode], k: int) -> bool:
        left,right = [],[] 
        lptr,rptr = root,root

        while lptr: 
            left.append(lptr)
            lptr=lptr.left 
        
        while rptr: 
            right.append(rptr)
            rptr=rptr.right

        while left and right: 
            l,r = left[-1].val,right[-1].val
            if l+r==k and not(l is r): 
                return True 
            if l+r>k: 
                right=self.prevItr(right)
            else: 
                left=self.nextItr(left) 
        
        return False 